### ISSUE TITLE
libvncserver 0.9.12: WebSocket unsolicited Pong (IE 11)

### ISSUE THREAD

**Issue by tbetker-rs** [LibVNC/libvncserver#337]
We are running _noVNC_ as VNC client in a browser window, with _x11vnc_ as VNC server (embedded Linux) and _libvncserver_ 0.9.12.

**Describe the bug**

When _noVNC_ is running in Internet Explorer 11 (Windows 7), the browser will send an unsolicited WebSocket Pong frame after ~30s of inactivity: `0x8A 0x80 .. .. .. ..` (6 bytes), no payload.

_libvncserver_ reports an error "webSocketsDecodeHybi: unhandled opcode 10, b0: 8a, b1: 80" and closes the connection. _x11vnc exits_, and _noVNC_ shows an error status "Server disconnected (code: 1006)".

The user has to refresh the browser window, and re-enter the password. This is rather annoying, so we got several complaints.

**Expected behaviour**

_libvncserver_ should ignore the Pong, and the connection should stay open.

**Workaround**

The workaround we are currently using consists of two steps:

Step 1. Handle Pong in `webSocketsDecodeHybi()`: Ignore the frame (including payload, if any), and return 0.

```
        memcpy(dst, payload, flength);
        result = flength;
        break;
+      case WS_OPCODE_PONG:
+       result = 0;
+       break;
       default:
        rfbErr("%s: unhandled opcode %d, b0: %02x, b1: %02x\n", __func__, (int)opcode, header->b0, header->b1);
     }
```

Step 2. Handle `webSocketsDecodeHybi()` returning 0: At the start of `rfbAuthProcessClientMessage()` and `rfbProcessClientNormalMessage()`, return gracefully when `rfbReadExact() == 0`, i.e., when no payload data is available.

```
     if ((n = rfbReadExact(cl, (char *)&msg, 1)) <= 0) {
-        if (n != 0)
+        if (n != 0) {
             rfbLogPerror("rfbProcessClientNormalMessage: read");
-        rfbCloseClient(cl);
+            rfbCloseClient(cl);
+        }
         return;
     }

```

This handles the cases were the client is in the states RFB_AUTHENTICATION (_x11vnc_ waits for password) or RFB_NORMAL (_x11vnc_ running), which is good enough for our purposes.

I did not modify `rfbProcessClientProtocolVersion()` (state RFB_PROTOCOL_VERSION), `rfbProcessClientSecurityType()` (RFB_SECURITY_TYPE), `rfbProcessClientInitMessage()` (RFB_INITIALISATION, RFB_INITIALISATION_SHARED), although these functions should probably be considered as well.

I am aware that this workaround may have unexpected side-effects. `n == 0` may indicate some errors other than "no payload data", and these errors are now ignored. In our setup, the workaround seems to be okay, though (i.e., works for me).

**Comment by tbetker-rs** [comment]
The same issue occurs with Edge on Windows 10 (EdgeHTML, not Chromium), as it sends Pongs as well.

**Comment by bk138** [comment]
As IE11 is deprecated by MS, I tentatively close this issue. I hope you're able to run a patched version of LibVNCServer for your specific use case.

**Comment by tbetker-rs** [comment]
Oh my, I completely forgot about this one. My fault.

Returning 0 in `webSocketsDecodeHybi()` was not such a good idea, of course, as I found out later; it means that the connection was closed, and we don't want to ignore that.

The correct reaction is to return -1, errno = EAGAIN; this error condition needs to be passed up to the callers so that `rfbAuthProcessClientMessage()` and `rfbProcessClientNormalMessage()` do nothing when no data is available.

For the record, I am attaching the patch I am currently using; it was written for libvncserver-0.9.10, and may have to be adapted for later versions (e.g., websockets.c code moved to ws_decode.c):
* In `webSocketsDecodeHybi()`, handle WS_OPCODE_PONG: Drop the frame, including the payload (if any), and return -1, errno = EAGAIN.
* In `rfbReadExactTimeout()`, set errno = EAGAIN when no data was read (and the timeout is 0).
* At the start of  `rfbAuthProcessClientMessage()` [state AUTHENTICATION] and `rfbProcessClientNormalMessage()` [NORMAL], accept that no data may be available.

For the last part, I am using a wrapper `rfbReadExact1()` which first tries to read a single byte with timeout 0, then the remaining bytes with the expected timeout:
* The wrapper returns -1, errno = EAGAIN when the first data byte is unavailable.
* Otherwise, it behaves the same as `rfbReadExact()`; in particular, it returns 0 when the peer has disconnected.
* The wrapper may still return -1, errno = ETIMEDOUT when at least one byte was read, but less than the exact length requested.

All this may probably be handled better, but it works for me.

I get it if you don't want to adopt the patch, given that IE 11 is deprecated (I am not sure about EdgeHTML). However, I do think that it would make sense to have some kind of "no data received" path, if only you want to support your own Ping function some day.

[pong.patch.zip](https://github.com/LibVNC/libvncserver/files/8599390/pong.patch.zip)
